src/dotnet/MLSearch.Service/Indexing/ChatIndexerWorker.cs

```diff
@@ -1,29 +1,69 @@
-using ActualChat.MLSearch.ApiAdapters;
-
 
 namespace ActualChat.MLSearch.Indexing;
 
 
 internal class ChatIndexerWorker(
     IDataIndexer<ChatId> dataIndexer,
-    ILoggerSource loggerSource
+    ILogger<ChatIndexerWorker> log
 ) : IChatIndexerWorker
 {
+    private record class Job(Task Completion, CancellationTokenSource Cancellation);
     private const int ChannelCapacity = 10;
-    private ILogger? _log;
-    private ILogger Log => _log ??= loggerSource.GetLogger(GetType());
 
     private readonly Channel<MLSearch_TriggerChatIndexing> _channel =
-        Channel.CreateBounded<MLSearch_TriggerChatIndexing>(new BoundedChannelOptions(ChannelCapacity){
+        Channel.CreateBounded<MLSearch_TriggerChatIndexing>(new BoundedChannelOptions(ChannelCapacity) {
             FullMode = BoundedChannelFullMode.Wait,
-            SingleReader = false,
+            SingleReader = true,
             SingleWriter = false,
         });
+    private readonly ConcurrentDictionary<ChatId, Job?> _runningJobs = new(-1, ChannelCapacity);
+    private readonly SemaphoreSlim _semaphore = new(ChannelCapacity, ChannelCapacity);
 
-    public ChannelWriter<MLSearch_TriggerChatIndexing> Trigger => _channel.Writer;
+    public async ValueTask PostAsync(MLSearch_TriggerChatIndexing input, CancellationToken cancellationToken)
+    {
+        await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I really-really don't want to acquire semaphore in one method and release it in another. It adds unnecessary complexity to manage and read

Reply:
As you can see, we acquire semaphore by the dispatch flow but release it by worker flows. So it can't be done in the same method.